File: studyOC/studyOC/Scheme和Universal/SceneDelegate+Scheme.m
//
//  SceneDelegate+Scheme.m
//  studyOC
//
//  Created by Hahn on 2021/9/15.
//

#import "SceneDelegate+Scheme.h"
#import "NSURL+Ex.h"
#import "LinkModel.h"

@implementation SceneDelegate (Scheme)

// 冷启动调用
- (void)scheme:(UIScene *)scene willConnectToSession:(UISceneSession *)session options:(UISceneConnectionOptions *)connectionOptions
{
    // 冷启动的时候，需要等初始化完成
    for (UIOpenURLContext *context in connectionOptions.URLContexts) {
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.50 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            NSString *url = context.URL.host;
            NSDictionary *paras = [context.URL getURLParameters];
            NSLog(@"SceneDelegate scheme url = %@",url);
            NSLog(@"SceneDelegate scheme 参数 = %@", paras);
            
            LinkModel *model = [LinkModel new];
            model.url = url;
            model.para = paras;
            [[NSNotificationCenter defaultCenter] postNotificationName:@"Scheme" object:model];
        });
    }
}

// 热启动调用
- (void)scene:(UIScene *)scene openURLContexts:(NSSet<UIOpenURLContext *> *)URLContexts
{
    NSString *url = URLContexts.allObjects.firstObject.URL.host;
    NSDictionary *paras = [URLContexts.allObjects.firstObject.URL getURLParameters];
    NSLog(@"SceneDelegate scheme url = %@",url);
    NSLog(@"SceneDelegate scheme 参数 = %@", paras);
    
    LinkModel *model = [LinkModel new];
    model.url = url;
    model.para = paras;
    [[NSNotificationCenter defaultCenter] postNotificationName:@"Scheme" object:model];
}

@end

File: studySwift/studySwift/Scheme和Universal/SceneDelegate+Scheme.swift
//
//  SceneDelegate+Scheme.swift
//  studySwift
//
//  Created by Hahn on 2021/9/16.
//

import UIKit

extension SceneDelegate {
    
    // 冷启动调用
    func scheme(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {
        guard let _ = (scene as? UIWindowScene) else { return }
        
        
        for context in connectionOptions.urlContexts{
 
            let hostUrl = context.url.host
            let paras = context.url.absoluteURL.urlStringtoParams(context.url.absoluteString)
            print("SceneDelegate scheme url = \(hostUrl)")
            print("SceneDelegate scheme paras = \(paras)")
            
            var model = LinkModel()
            var para = Para()
            model.url = hostUrl
            para.name = paras["name"] as? String
            para.id = paras["id"] as? String
            para.age = paras["age"] as? String
            model.para = para
            
            DispatchQueue.main.asyncAfter(deadline: .now()+0.5, execute:
                                            {
                                                NotificationCenter.default.post(name: NSNotification.Name.init(rawValue: "Scheme"), object: model)
                                            })
                
            
        }
        
        
        
    }
    
    // 热启动调用
    func scene(_ scene:UIScene, openURLContexts URLContexts:Set<UIOpenURLContext>) {
        guard let url = URLContexts.first?.url else { return }
        let hostUrl = url.host
        let paras = url.absoluteURL.urlStringtoParams(url.absoluteString)
        print("SceneDelegate scheme url = \(hostUrl)")
        print("SceneDelegate scheme paras = \(paras)")
        
        var model = LinkModel()
        var para = Para()
        model.url = hostUrl
        para.name = paras["name"] as? String
        para.id = paras["id"] as? String
        para.age = paras["age"] as? String
        model.para = para
        
        NotificationCenter.default.post(name: NSNotification.Name.init(rawValue: "Scheme"), object: model)
    }
    
}
